fdbserver/TLogServer.actor.cpp

```diff
@@ -1513,12 +1513,15 @@ std::deque<std::pair<Version, LengthPrefixedStringRef>>& getVersionMessages(Refe
 	return tagData->versionMessages;
 };
 
-Future<Void> waitForMessagesForTag(Reference<LogData> self, TLogPeekRequest const& req) {
+ACTOR Future<Void> waitForMessagesForTag(Reference<LogData> self, TLogPeekRequest const& req) {
 	auto tagData = self->getTagData(req.tag);
 	if (tagData.isValid()) {
 		return Void();
 	}
-	return self->waitingTags[req.tag].getFuture();
+	wait(self->waitingTags[req.tag].getFuture());
+	// we want the caller to finish first, otherwise the data structure it is building might not be complete
```

Comment:
I assume tlog is single-threaded, so I don't see the underlying issue. Maybe we need to deeper understanding of flow to understand the issue here.

Reply:
`promise.send(...)` will immediately call into all waiting futures. So the caller needs to make sure that all invariants are holding before calling `send`. the `wait(delay(0))` simply relaxes this requirement.